spec/lib/whitehall_importer/create_image_revision_spec.rb

```diff
@@ -0,0 +1,110 @@
+# frozen_string_literal: true
+
+RSpec.describe WhitehallImporter::CreateImageRevision do
+  describe "#call" do
+    let(:img_url) { "https://assets.publishing.service.gov.uk/government/uploads/valid-image.jpg" }
+    let(:img_file) { "960x640.jpg" }
+    let(:whitehall_image) do
+      { "url": img_url, "caption": "Caption", "alt_text": "Alt text" }.with_indifferent_access
+    end
+
+    before do
+      stub_request(:get, img_url).to_return(
+        status: 200,
+        body: file_fixture(img_file).read,
+      )
+    end
+
+    context "A valid image is provided" do
+      before do
+        stub_request(:get, img_url).to_return(status: 200, body: file_fixture("960x640.jpg"))
+      end
+
+      it "should create an Image::MetadataRevision that preserves the alt text and caption" do
+        described_class.call(whitehall_image)
+        expect(Image::MetadataRevision.count).to eq(1)
+        expect(Image::MetadataRevision.first.caption).to eq(whitehall_image["caption"])
+        expect(Image::MetadataRevision.first.alt_text).to eq(whitehall_image["alt_text"])
+      end
+
+      it "should create an Image::BlobRevision that preserves the filename" do
+        described_class.call(whitehall_image)
+        expect(Image::BlobRevision.count).to eq(1)
+        expect(Image::BlobRevision.first.filename).to eq("valid-image.jpg")
+      end
+
+      it "should create an Image::Revision that references the Metadata and Blob revisions" do
+        described_class.call(whitehall_image)
+        expect(Image::Revision.count).to eq(1)
+        expect(Image::Revision.first.blob_revision).to eq(Image::BlobRevision.first)
+        expect(Image::Revision.first.metadata_revision).to eq(Image::MetadataRevision.first)
+      end
+    end
+
+    context "Image not available" do
+      before do
+        stub_request(:get, img_url).to_return(status: 404)
+      end
+
+      it "should raise a WhitehallImporter::AbortImportError" do
+        expect { described_class.call(whitehall_image) }.to raise_error(
+          WhitehallImporter::AbortImportError,
+          "Image does not exist: #{img_url}",
+        )
+      end
+
+      it "should not create any revision instances" do
+        begin
+          described_class.call(whitehall_image)
+        rescue WhitehallImporter::AbortImportError
+          expect(Image::MetadataRevision.count).to eq(0)
+          expect(Image::BlobRevision.count).to eq(0)
+          expect(Image::Revision.count).to eq(0)
+        end
```

Comment:
I don't think we're too concerned about whether these are created or not. We expect AbortImportError to break a transaction and that will involve some data being persisted and discarded. So although I can't see a situation where we'd expect these to actually exist I also don't think we have reason to care about whether any were or not, just the exception being raised should be sufficient.

Reply:
Agreed that this test can be removed, as really we would want one for every bit of the import that creates something, not just the image logic

I do think it's worth having a higher level test that ensures we're wrapping the entire import in a Transaction - one for a small future PR?